Torch deploy for fx.grapm_module with non-torch dependencie

HarutMov opened the pull request:
Summary:
This diff enables torch deploy for fx.graph_module with non-torch dependencies . Here are the issues currently preventing this and are fixed in this change:
-  Pickle is used as an internal format to transmit objects between interpreters. It needs to serialize python code, but to be able to get the source code for imports from python_code.globals it needs access to the PackageImporter. Currently a regular _reduce_ function is used which doesn't have the notion of custom importer.
- When deserializing pickled objects on an interpreter, it is passing empty globals to exec, thus it will not be able to resolve non-torch imports located in the package. We need to be able to point exec to our custom PackageImporter.
- Subclasses extending fx.graph_module should be able to optionally provide their own Tracer (extending fx.Tracer).

As a solution a new reducer is introduced (_reduce_deploy_) for torch deploy workflow. Reducer will be registered in _deploy.py (entry point for C++ torch deploy API) when saving the object transmitting it between interpreters. It allows us to pass a proper PackageImporter for each interpreter for pickling/unpickling fx.graph_module. It also defines an api for passing custom fx.Tracer when needed.

Test Plan:
Added UT to cover changes.
```
buck test //caffe2/torch/csrc/deploy:test_deploy
```
```
buck test caffe2/test:fx
```

Differential Revision: D29690088

suo commented on torch/fx/graph_module.py:
```diff
@@ -515,6 +533,15 @@ def wrapped_call(self, *args, **kwargs):
 
+    # Passing Tracer as argument allows subclasses extending fx.GraphModule
+    # define their own Tracer (extending fx.Tracer).
+    def __reduce_deploy__(self, importer: PackageImporter, clsTracer: Any = None):
```
What's the motivation behind making `clsTracer` an `Any` as opposed to `Tracer`?

HarutMov replied:
We need to pass a class object, not the actual Tracer object . class object argument allows subclasses extending fx.GraphModule define their own Tracer (used in _deserialize_graph_module).

For moving to Type[Tracer] instead of Any, we would need to import Tracer module globally here, but it would introduce a circular dependency. Thats why in _deserialize_graph_module it is imported before using.

The best we can do here IMO is to move from Any to Type.

suo commented on torch/fx/graph_module.py:
```diff
@@ -68,17 +68,29 @@ def reduce_graph_module(body: Dict[Any, Any], import_block: str) -> torch.nn.Mod
+    return _deserialize_graph_module(forward, body, clsTracer)
+
+
+def _deserialize_graph_module(forward, body: Dict[Any, Any], clsTracer: Any = None) -> torch.nn.Module:
```
This change to `_deserialize_graph_module` is not backwards compatible. Old serialized GraphModules will continue to call this function, so we need to make sure that they continue to work.

HarutMov replied:
As was discuss offline _deserialize_graph_module function signature is not being packaged, so this will not break BC.

suo commented on torch/fx/graph_module.py:
```diff
@@ -68,17 +68,29 @@ def reduce_graph_module(body: Dict[Any, Any], import_block: str) -> torch.nn.Mod
+def reduce_deploy_graph_module(importer: PackageImporter,
+                               body: Dict[Any, Any],
+                               import_block: str,
+                               clsTracer: Any) -> torch.nn.Module:
```
Also this is super dumb, but I feel like it should be `tracer_cls` instead of `cls_tracer` everywhere :P

HarutMov replied:
renamed per suggestion

HarutMov committed:
Torch deploy for fx.grapm_module with non-torch dependencie (#61680)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61680

This diff enables torch deploy for fx.graph_module with non-torch dependencies . Here are the issues currently preventing this and are fixed in this change:
-  Pickle is used as an internal format to transmit objects between interpreters. It needs to serialize python code, but to be able to get the source code for imports from python_code.globals it needs access to the PackageImporter. Currently a regular _reduce_ function is used which doesn't have the notion of custom importer.
- When deserializing pickled objects on an interpreter, it is passing empty globals to exec, thus it will not be able to resolve non-torch imports located in the package. We need to be able to point exec to our custom PackageImporter.
- Subclasses extending fx.graph_module should be able to optionally provide their own Tracer (extending fx.Tracer).

As a solution a new reducer is introduced (_reduce_deploy_) for torch deploy workflow. Reducer will be registered in _deploy.py (entry point for C++ torch deploy API) when saving the object transmitting it between interpreters. It allows us to pass a proper PackageImporter for each interpreter for pickling/unpickling fx.graph_module. It also defines an api for passing custom fx.Tracer when needed.

Test Plan:
Added UT to cover changes.
```
buck test //caffe2/torch/csrc/deploy:test_deploy
```
```
buck test caffe2/test:fx
```

Differential Revision: D29690088

fbshipit-source-id: c8f9c762ce1d0a2ec960445cdadecde468464cec

suo reviewed (approved):
lgtm!
